src/Interpreters/Cache/FileCache.cpp

```diff
@@ -764,14 +764,16 @@ FileCache::getOrSet(
         }
     }
 
+    /// Compare with initial_range and not result_range,
+    /// See comment in splitRange for explanation.
     chassert(file_segments_limit
-             ? file_segments.back()->range().left <= result_range.right
-             : file_segments.back()->range().contains(result_range.right),
+             ? file_segments.back()->range().left <= initial_range.right
```

Comment:
This assert was hitting exactly when the case described here was reproduced https://github.com/ClickHouse/ClickHouse/blob/bbbf612e19a8c391d0f4185bb739d1f182a6e651/src/Interpreters/Cache/FileCache.cpp#L403-L424